<div align='center'>
МІНІСТЕРСТВО ОСВІТИ І НАУКИ УКРАЇНИ

Національний аерокосмічний університет ім. М.Є.Жуковського
"Харківський авіаційний інститут"

Факультет радіотехніки, комп’ютерних систем і інфокомунікацій

Кафедра інформаційно-комунікаційних технологій ім. О. О. Зеленського


ПРАКТИЧНА РОБОТА №3

з курсу "Захист інформації в інфокомунікаціях"

на тему "Перевірка активності кінцевого пристрою"

<div align='right'>
Виконав: студент ??? курсу

групи № ???

спеціальності 172 "Телекомунікації та радіотехника" / 126 "Інформаційні системи і технології" (залишити необхідне)

Прізвище та ініціали

Прийняв: (вказати посаду)

Прізвище та ініціали викладача

Національна шкала: __________

Кількість балів: _____

Оцінка: ECTS _____ 
</div>

Харків - 2023
</div>

## Практична робота №3. Перевірка активності кінцевого пристрою

**Мета роботи**: ознайомитися з синтаксисом мови програмування Python і бібліотеки Scapy, ознайомитися зі структурою заголовків мережних протоколів, скласти скрипт, що відправляє запити за протоколом ICMP, з метою виявлення пристроїв у мережі.

### Зміст

- Частина 1. Команда ping та ICMP-запити
- Частина 2. Підготовка Scapy
- Частина 3. Scapy і заголовки мережних протоколів
- Частина 4. ICMP-запити засобами Scapy

### Частина 1. Команда ping та ICMP-запити

**Завдання** - перевірка зв'язку з заданим вузлом командою ping

Для перевірки зв'язку переважно використовують системну команду `ping`, що надсилає ICMP-пакети на задану адресу і отримує відповідь про досяжність чи недосяжність вказаного вузла

**Крок 1. Обрати адресу призначення за варіантом по списку**

|Варіант|Сайт|Кількість запитів N|
|---|---|---|
|1|google.com|6|
|2|yahoo.com|3|
|3|facebook.com|4|
|4|python.org|5|
|5|youtube.com|6|
|6|wikipedia.org|3|
|7|amazon.com|4|
|8|live.com|5|
|9|bing.com|6|
|10|instagram.com|3|
|11|stackoverflow.com|4|
|12|whatsapp.com|5|
|13|twitter.com|6|
|14|telegram.com|3|
|15|viber.com|4|
|16|ukr.net|5|
|17|accuweather.com|6|
|18|sinoptik.ua|3|
|19|gismeteo.ua|4|
|20|i.ua|5|
|21|www.netacad.com|6|

**Крок 2. Створити ping-запит** \
Для заданого **сайту** виконати запит командою `ping` з відправкою **N** пакетів (у Windows використовуйте параметр `-n`, у Linux - `-c`)

In [8]:
!chcp 65001

Active code page: 65001


In [9]:
!ping google.com


Pinging google.com [142.250.203.206] with 32 bytes of data:
Reply from 142.250.203.206: bytes=32 time=24ms TTL=119
Reply from 142.250.203.206: bytes=32 time=24ms TTL=119
Reply from 142.250.203.206: bytes=32 time=24ms TTL=119
Reply from 142.250.203.206: bytes=32 time=24ms TTL=119

Ping statistics for 142.250.203.206:
    Packets: Sent = 4, Received = 4, Lost = 0 (0% loss),
Approximate round trip times in milli-seconds:
    Minimum = 24ms, Maximum = 24ms, Average = 24ms


**Крок 3. Визначити в мережевих налаштуваннях ПК мережну адресу пристрою і мережі в цілому** \
Команди командного рядка(терміналу) для визначення налаштувань мережі: 
- для Windows - `ipconfig /all` (всі налаштування разом)
- для Linux - `ip addr` для визначення мережних адрес пристрою та `ip route` для визначення шлюза за замовчуванням
- або зайти в графічний інтерфейс налаштувань мережі

In [10]:
!ip addr

1: lo: <LOOPBACK,UP,LOWER_UP> mtu 65536 qdisc noqueue state UNKNOWN group default qlen 1000
    link/loopback 00:00:00:00:00:00 brd 00:00:00:00:00:00
    inet 127.0.0.1/8 scope host lo
       valid_lft forever preferred_lft forever
    inet6 ::1/128 scope host 
       valid_lft forever preferred_lft forever
2: enp61s0: <NO-CARRIER,BROADCAST,MULTICAST,UP> mtu 1500 qdisc mq state DOWN group default qlen 1000
    link/ether 2c:f0:5d:65:a2:52 brd ff:ff:ff:ff:ff:ff
3: wlo1: <BROADCAST,MULTICAST,UP,LOWER_UP> mtu 1500 qdisc noqueue state UP group default qlen 1000
    link/ether f8:ac:65:88:a7:2a brd ff:ff:ff:ff:ff:ff
    altname wlp0s20f3
    inet 192.168.31.195/24 brd 192.168.31.255 scope global dynamic noprefixroute wlo1
       valid_lft 35823sec preferred_lft 35823sec
    inet6 fe80::a28d:4cc9:8eb7:4f88/64 scope link noprefixroute 
       valid_lft forever preferred_lft forever


**Крок 4. Відповісти на питання**
1. Яка ІР-адреса сайту за результатами виконання команди ping? __________
2. Яка ІР-адреса поточного пристрою? ____________
3. Яка МАС-адреса поточного пристрою? ____________
4. Яка ІРv6-адреса поточного пристрою? ____________
5. Яка ІР-адреса шлюзу за замовчуванням? ____________

**Крок 5. Привести в звіті результати роботи Ping**

Навести в звіті результати виконання кроків №2-4 \
*Не потрібно у разі використання даного шаблону jupyter-notebook*

### Частина 2. Підготовка Scapy

Підготовка Jupyter-notebook i Scapy для роботи з мережею

**Крок 1. Встановлення Scapy**

Scapy - бібліотека python для роботи з мережними пакетами та протоколами. \
Документація: <https://scapy.readthedocs.io/en/latest/index.html>

1. Вимога - попередньо мають бути встановлені `python3` і пакетний менеджер `pip`
2. В командному рядку виконати команду: \
`pip install scapy`

**Особливості**:
- **Windows**. У випадку попередження \
`Warning: No libpcap provider available! pcap won't be used` \
слід також встановити програму-аналізатор мережних пакетів Wireshark з офіційної сторінки <https://www.wireshark.org>
- **Linux** - програма потребує встановлення і доступу з правами адміністратора: \
`sudo pip install notebook scapy`

**Крок 2. Запуск Scapy з правами адміністратора**

1. Запуск jupyter-notebook з правами адміністратора як середовища для виконання всіх подальших запитів
- **Linux** - найпростіший варіант - запуск jupyter-notebook з правами адміністратора:
    - Запустити термінал в окремому вікні або як вкладку середовища розробки(не буде працювати у випадку встановлення IDE як Snap/Flatpak пакет)
    - Запустити з правами адміністратора: \
    ```
    sudo jupyter notebook --allow-root
    ```
- **Windows** - *попередньо не потрібно, на стадії тестування*
    - За необхідності найпростіший варіант - в "Проводнику" через пошук знайти `jupyter-notebook.exe` і через контекстне меню обрати "як Адміністратор"

1. Підключення до запущеного сервера jupyter-notebook - після запуску в терміналі(командному рядку) відображається посилання для доступу до Jupyter-сервера в форматі
   `http://127.0.0.1:8888/tree?token=0c86a75d0aa67fc4cb412ee587b990806b3095b63e9d1a6f`
 
- **Робота в браузері** - відкрити посилання і знайти файл блокнота в домашньому каталозі
- **Робота в Intellij IDEA / PyCharm**: Tools -> Add Jupyter Connection -> Connect to Jupyter server using URL (не запрацювало)
- **Робота в Visual Studio Code**: Select Kernel -> Select Another Kernel -> Existing Jupyter Server -> Enter the URL` -> Ввести URL, підтвердити назву і kernel:Python3

**Крок 3. Імпорт бібліотеки Scapy в Python**

In [2]:
from scapy.all import *

### Частина 3. Scapy і заголовки мережних протоколів


**Завдання** - перевірка зв'язку з заданим вузлом, потребує реалізації завдання з частини №1 за допомогою Python та Scapy

**Крок 1. Ознайомитися з документацією і прикладами Scapy**

Документація:
- Офіційна документація з прикладами - <https://scapy.readthedocs.io/en/latest/usage.html>
- Блокнот з перекладеними прикладами - <https://github.com/OlegIeremeiev/InfoSecurity/blob/main/Docs_scapy_general.ipynb>


**Крок 2. Створити ICMP-пакет для заданого сайту**

Для заданого **сайту** та кількості запитів **N** створити ICMP-пакет, використовуючи можливості Python та Scapy.

In [36]:
ping = Ether()/

**Крок 3. Відправити і отримати відповіді для заданої кількості пакетів** \
Для відправки пакетів і отримання відповіді використовуються функції `sr`, `sr1`, `srp`, `srp1` (докладніше в документації).

In [ ]:
answered, unanswered = 

**Крок 4. Отримати інформацію з відповіді** \
Змінна `answered` представляє собою масив запитів і відповідей у форматі:
- `answered[номер запиту][0]` - пакет запиту
- `answered[номер запиту][1]` - пакет відповіді

В кожному пакеті доступ до поля заголовку одного з протоколів забезпечується по його імені подібно до словника (в залежності від типу пакета і наявних протоколів):
- `packet[Ether]`
- `packet[IP]`
- `packet[ICMP]`
- `packet[TCP]`
- `packet[UDP]`
- тощо

1. Вивести загальну інформацію відповіді функцією `show()`

2. Вивести тільки МАС-адресу відправника запиту

3. МАС-адреса якого пристрою вказана як МАС-адреса отримувача запиту? ___________

4. Вивести тільки ІР-адреси відправника запиту і отримувача запиту:

**Крок 5. Навести в звіті результати роботи завдання**

Навести в звіті результати виконання кроку №4 \
*Не потрібно у разі використання даного шаблону jupyter-notebook*

### Частина 4. ICMP-запити засобами Scapy

**Завдання** - реалізація команди ping засобами Scapy. В даному завданні необхідно навчитися використовувати функцію `sniff()` для захоплення всього трафіку, що проходить через мережний інтерфейс і реалізувати власну функцію `ping`

**Крок 1. Захоплення пакетів сторонніх програм** \
Функція `sniff()` працює безкінечно або завершує роботу після захоплення вказаної кількості пакетів з можливістю фільтрації по типу пакетів

In [14]:
capture = sniff(iface="wlo1", filter="icmp", count=10)
capture.show()

0000 Ether / IP / ICMP 192.168.31.195 > 142.250.203.142 echo-request 0 / Raw
0001 Ether / IP / ICMP 142.250.203.142 > 192.168.31.195 echo-reply 0 / Raw
0002 Ether / IP / ICMP 192.168.31.195 > 142.250.203.142 echo-request 0 / Raw
0003 Ether / IP / ICMP 142.250.203.142 > 192.168.31.195 echo-reply 0 / Raw
0004 Ether / IP / ICMP 192.168.31.195 > 142.250.203.142 echo-request 0 / Raw
0005 Ether / IP / ICMP 142.250.203.142 > 192.168.31.195 echo-reply 0 / Raw
0006 Ether / IP / ICMP 192.168.31.195 > 142.250.203.142 echo-request 0 / Raw
0007 Ether / IP / ICMP 142.250.203.142 > 192.168.31.195 echo-reply 0 / Raw
0008 Ether / IP / ICMP 192.168.31.195 > 142.250.203.142 echo-request 0 / Raw
0009 Ether / IP / ICMP 142.250.203.142 > 192.168.31.195 echo-reply 0 / Raw


**Крок 2. Форматування результатів** \
Необхідно написати код, що дозволяє форматувати інформацію запиту подібно до результатів роботи команди `ping`


**Стиль Windows**: \
\> C:\Users\oleg>ping google.com \
~~Обмен пакетами с google.com [172.217.16.46] с 32 байтами данных:~~ \
Ответ от 172.217.16.46: число байт=32 время=23мс TTL=127 \
Ответ от 172.217.16.46: число байт=32 время=25мс TTL=127 \
Ответ от 172.217.16.46: число байт=32 время=24мс TTL=127 \
Ответ от 172.217.16.46: число байт=32 время=24мс TTL=127

Зафіксувати час можно використовуючи `TimeStamp` пакетів - `p[ICMP].time` \
Перелік всіх полів окремого пакета можно побачити функцією `ls()`

In [27]:
ls(capture[1])

dst        : DestMACField                        = 'f8:ac:65:88:a7:2a' ('None')
src        : SourceMACField                      = '9c:9d:7e:49:4a:8e' ('None')
type       : XShortEnumField                     = 2048            ('36864')
--
version    : BitField  (4 bits)                  = 4               ('4')
ihl        : BitField  (4 bits)                  = 5               ('None')
tos        : XByteField                          = 0               ('0')
len        : ShortField                          = 84              ('None')
id         : ShortField                          = 0               ('1')
flags      : FlagsField                          = <Flag 0 ()>     ('<Flag 0 ()>')
frag       : BitField  (13 bits)                 = 0               ('0')
ttl        : ByteField                           = 119             ('64')
proto      : ByteEnumField                       = 1               ('0')
chksum     : XShortField                         = 2229            ('None')
src       

Windows-style(приклад): \
Ответ от 142.250.203.142: число байт=64 время=25мс TTL=119 \
Ответ от 142.250.203.142: число байт=64 время=26мс TTL=119 \
Ответ от 142.250.203.142: число байт=64 время=25мс TTL=119 \
Ответ от 142.250.203.142: число байт=64 время=26мс TTL=119 \
Ответ от 142.250.203.142: число байт=64 время=25мс TTL=119

**Бонусне завдання - стиль Linux**: \
\$ ping google.com \
PING google.com (142.250.203.142) 56(84) bytes of data. \
64 bytes from waw07s06-in-f14.1e100.net (142.250.203.142): icmp_seq=1 ttl=119 time=25.9 ms \
64 bytes from waw07s06-in-f14.1e100.net (142.250.203.142): icmp_seq=2 ttl=119 time=25.6 ms \
64 bytes from waw07s06-in-f14.1e100.net (142.250.203.142): icmp_seq=3 ttl=119 time=24.9 ms \
64 bytes from waw07s06-in-f14.1e100.net (142.250.203.142): icmp_seq=4 ttl=119 time=25.5 ms \

З коду: \
64 bytes from waw07s06-in-f14.1e100.net (142.250.203.142): icmp_seq=0 ttl=119 time=25.0 ms \
64 bytes from waw07s06-in-f14.1e100.net (142.250.203.142): icmp_seq=1 ttl=119 time=26.0 ms \
...

**Крок 3. Застосування Asynchronous Sniffing**

Данний режим дозволяє запускати і завершувати сніфінг за потреби після виконання необхідного коду

```python
t = AsyncSniffer()
t.start()
print("hey") # hey
...
results = t.stop()
```

Використовуючи даний режим зробіть запити з форматованим відображенням результатів

In [77]:
t = AsyncSniffer(iface="wlo1", filter="icmp")
t.start()

# place for pings



capture = t.stop()
capture.show()

Begin emission:
Finished sending 5 packets.
.**...............*.........*.*
Received 31 packets, got 5 answers, remaining 0 packets
0000 Ether / IP / ICMP 192.168.31.195 > 142.250.203.142 echo-request 0
0001 Ether / IP / ICMP 142.250.203.142 > 192.168.31.195 echo-reply 0
0002 Ether / IP / ICMP 192.168.31.195 > 142.250.203.142 echo-request 0
0003 Ether / IP / ICMP 142.250.203.142 > 192.168.31.195 echo-reply 0
0004 Ether / IP / ICMP 192.168.31.195 > 142.250.203.142 echo-request 0
0005 Ether / IP / ICMP 142.250.203.142 > 192.168.31.195 echo-reply 0
0006 Ether / IP / ICMP 192.168.31.195 > 142.250.203.142 echo-request 0
0007 Ether / IP / ICMP 142.250.203.142 > 192.168.31.195 echo-reply 0
0008 Ether / IP / ICMP 192.168.31.195 > 142.250.203.142 echo-request 0
0009 Ether / IP / ICMP 142.250.203.142 > 192.168.31.195 echo-reply 0


**Крок 4. Навести в звіті результати роботи завдання**

Навести в звіті результати виконання кроків №2-3 \
*Не потрібно у разі використання даного шаблону jupyter-notebook*